### PR TITLE
Add data version to framework and chip-tool-darwin.

### LINKED ISSUE #16136: TC-IDM testing: Chip-tool-darwin needs to support the"Dataversion field" and "SuppressResponse" value to set.
As per steps in testcases [TC-IDM-2.2][TC-IDM-3.2] and [TC-IDM-4.2], chip-tool-darwin needs support for "Dataversion field" and as per steps in testcases [TC-IDM-1.2], [TC-IDM-2.1], [TC-IDM-3.1], chip-tool-darwin needs support for "SuppressResponse field".

### PR DESCRIPTION
**Pull request by krypton36**
#### Problem
CHIP.framework has no way of setting the data version in read and write. We need a way to set this in the framework and in chip-tool-darwin.
* Fixes #16136 

#### Change overview
- Updates CHIP.framework Read APIs to include params and retains legacy API.
- Updates CHIP.framework ReadParams to have dataVersion and sets it.
- Updates CHIP.framework WriteParams to have dataVersion and sets it.
- Updates chip-tool-darwin to accept data version as an argument and sets it.

#### Testing
- Tested legacy read
- Tested new data version read and write with accessory.

### CODE REVIEW

File: examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h [hunk 1]
```diff
@@ -44,6 +44,7 @@ class ReadAttribute : public ModelCommand {
         : ModelCommand("read")
     {
         AddArgument("attr-name", attributeName);
+        AddArgument("data-version", 0, UINT32_MAX, &mDataVersion);
```
Comment (bzbarsky-apple) [inline examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h:47]: Shouldn't the read-by-id versions also allow this?

File: examples/chip-tool-darwin/commands/clusters/WriteAttributeCommandBridge.h [hunk 2]
```diff
@@ -47,6 +47,7 @@ class WriteAttribute : public ModelCommand {
     WriteAttribute(const char * _Nonnull attributeName)
         : ModelCommand("write")
     {
+        AddArgument("data-version", 0, UINT32_MAX, &mDataVersion);
```
Comment (bzbarsky-apple) [inline examples/chip-tool-darwin/commands/clusters/WriteAttributeCommandBridge.h:50]: Similarly, write-by-id should allow this?

File: src/darwin/Framework/CHIP/CHIPCluster.h [hunk 3]
```diff
@@ -57,6 +57,21 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ * This tag can be omitted if the value of EnableTagCompression in the Path field is true.
+ * In this case, the value for the omitted tag SHALL be set to the value for that
+ * tag (if present) in the last AttributeDataIB that had tag compression disabled
+ * and was seen in a message that is part of the same interaction model Action
+ * as the current message.
+ *
+ * If this tag was not present and tag compression was disabled,
+ * it SHALL be interpreted as though a data version was not specified in that,
+ * or subsequent AttributeDataIBs.
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPCluster.h:71]: This comment makes no sense as API documentation here.  What is "EnableTagCompression"?  That's a low-level concept this API does not know about.

How about:
```suggestion
 * If not nil, the write will only succeed if the current data version of the cluster matches the provided data version.
```

File: src/darwin/Framework/CHIP/CHIPCluster.h [hunk 4]
```diff
@@ -80,6 +95,21 @@ NS_ASSUME_NONNULL_BEGIN
...
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPCluster.h:109]: This does not make any sense, again.

But also, if the data version is not nil and matches the server-side data version... what is the actual behavior of the read?  I suspect it doesn't really work very well.  How was this part tested?

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt [hunk 5]
```diff
@@ -37,7 +37,7 @@ using namespace chip::app::Clusters;
 {{/unless}}
 - (void){{asLowerCamelCase name}}WithParams: (CHIP{{asUpperCamelCase parent.name}}Cluster{{asUpperCamelCase name}}Params * {{#unless (commandHasRequiredField .)}}_Nullable{{/unless}})params completionHandler:({{>command_completion_type command=.}})completionHandler
 {
-    chip::Optional<uint16_t> timedInvokeTimeoutMs;
+    chip::Optional<uint16_t> timedInvokeTimeoutMs = chip::NullOptional;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt:40]: Fwiw, NullOptional is what the default constructor constructs to.

File: examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h (hunk 1 above)
Comment (krypton36) [reply examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h:47]: I think it should also be included in the read-by-id, but the CHIP.framework does not support passing in this.

File: examples/chip-tool-darwin/commands/clusters/WriteAttributeCommandBridge.h (hunk 2 above)
Comment (krypton36) [reply examples/chip-tool-darwin/commands/clusters/WriteAttributeCommandBridge.h:50]: I think it should also be included in the write-by-id, but the CHIP.framework does not support passing in this.

File: src/darwin/Framework/CHIP/CHIPCluster.h (hunk 4 above)
Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPCluster.h:109]: I am not certain what the server side version is, but I can look into it and attempt it. I tried with version 0 and 1, the read was successful. Although with write, I kept getting incompatible version error from the server.

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt (hunk 5 above)
Comment (krypton36) [reply src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt:40]: Ok. Let me remove this. Thank you.

File: examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h:47]: OK.  Can you please make sure an issue is filed?

File: examples/chip-tool-darwin/commands/clusters/WriteAttributeCommandBridge.h (hunk 2 above)
Comment (bzbarsky-apple) [reply examples/chip-tool-darwin/commands/clusters/WriteAttributeCommandBridge.h:50]: Again, please file issues as needed.

File: examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h (hunk 1 above)
Comment (krypton36) [reply examples/chip-tool-darwin/commands/clusters/ReportCommandBridge.h:47]: https://github.com/project-chip/connectedhomeip/issues/18334

File: src/darwin/Framework/CHIP/CHIPCluster.h (hunk 3 above)
Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPCluster.h]: Updated.

File: src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt (hunk 5 above)
Comment (krypton36) [reply src/darwin/Framework/CHIP/templates/CHIPClustersObjc-src.zapt]: Updated

File: src/darwin/Framework/CHIP/CHIPCluster.h (hunk 4 above)
Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPCluster.h]: Updated comment.

Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPCluster.h]: Ok. I removed data version from read attribute and created an issue to add it back is OnDone callback is added [here](https://github.com/project-chip/connectedhomeip/issues/18339).

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
Fast-tracking platform-specific change.

**Pull request merged**